fe/fe-core/src/main/java/com/starrocks/sql/ast/ExpressionPartitionDesc.java

```diff
@@ -105,34 +128,53 @@ public void analyze(List<ColumnDef> columnDefs, Map<String, String> otherPropert
     }
 
     @Override
-    public PartitionInfo toPartitionInfo(List<Column> schema, Map<String, Long> partitionNameToId,
-                                         boolean isTemp, boolean isExprPartition)
+    public PartitionInfo toPartitionInfo(List<Column> schema, Map<String, Long> partitionNameToId, boolean isTemp)
             throws DdlException {
-        PartitionType partitionType = PartitionType.RANGE;
-        if (isExprPartition) {
-            partitionType = PartitionType.EXPR_RANGE;
-        }
-        // we will support other PartitionInto in the future
+        // for materialized view express partition.
         if (rangePartitionDesc == null) {
-            // for materialized view express partition.
-            return new ExpressionRangePartitionInfo(Collections.singletonList(expr), schema, partitionType);
+            return new ExpressionRangePartitionInfo(Collections.singletonList(expr), schema, PartitionType.RANGE);
         }
         List<Column> partitionColumns = Lists.newArrayList();
-
         // check and get partition column
         for (String colName : rangePartitionDesc.getPartitionColNames()) {
             findRangePartitionColumn(schema, partitionColumns, colName);
         }
+        // automatic partition / partition expr only support one partition column
+        Column sourcePartitionColumn = partitionColumns.get(0);
+        if (partitionType != null) {
+            Column newTypePartitionColumn = new Column(sourcePartitionColumn);
+            newTypePartitionColumn.setType(partitionType);
+            partitionColumns = Lists.newArrayList(newTypePartitionColumn);
+        }
+        for (Column column : partitionColumns) {
+            try {
+                RangePartitionInfo.checkRangeColumnType(column);
+            } catch (AnalysisException e) {
+                throw new DdlException(e.getMessage());
+            }
+        }
 
-        ExpressionRangePartitionInfo expressionRangePartitionInfo =
-                new ExpressionRangePartitionInfo(Collections.singletonList(expr), partitionColumns, partitionType);
+        // Recreate a partition column type bypass check
+        RangePartitionInfo partitionInfo;
+        if (rangePartitionDesc.isAutoPartitionTable) {
+            // for automatic partition table
+            partitionInfo = new ExpressionRangePartitionInfo(Collections.singletonList(expr), partitionColumns,
+                    PartitionType.EXPR_RANGE);
+        } else {
+            // for partition by range expr
+            ExprPartitionMeta meta = new ExprPartitionMeta();
+            meta.setAutomaticPartition(false);
+            meta.setPartitionExprs(Collections.singletonList(expr));
+            meta.setSourcePartitionTypes(Collections.singletonList(sourcePartitionColumn.getType()));
+            partitionInfo = new ExprRangePartitionInfo(meta, partitionColumns);
```

Comment:
now in which case to use ExprRangePartitionInfo? which case to use ExpressionRangePartitionInfo?